pymagicc/input.py

```diff
@@ -0,0 +1,225 @@
+        :param stream: A Streamlike object (nominally StringIO) containing the
+            table to be extracted
+        :param metadata: Dictionary containing
+        :return: Tuple of a pd.DataFrame containing the data and a Dict
+            containing the units for each gas present in the output. The
+            pd.DataFrame columns are named using a MultiIndex of Gas and Region
+        """
+        raise NotImplementedError()
+
+
+class MAGICC6Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gas = basename(self.fname).split('_')[1]
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            index_col=0,
+            engine="python")
+        # Convert to a columns to a MultiIndex
+        df.columns = [
+            [gas] * len(df.columns),
+            df.columns
+        ]
+        df.index.name = 'YEAR'
+
+        units = {
+            gas: metadata['units']
+        }
+        return df, units
+
+
+class MAGICC7Reader(InputReader):
+    def process_data(self, stream, metadata):
+        gases = stream.readline().split()
+        assert gases[0] == 'GAS'
+        gases = gases[1:]
+        stream.readline()  # Unused line
+        units = stream.readline().split()[1:]
+        regions = stream.readline().split()[1:]
+        index = pd.MultiIndex.from_arrays([gases, regions], names=['GAS', 'REGION'])
+        df = pd.read_csv(
+            stream,
+            skip_blank_lines=True,
+            delim_whitespace=True,
+            names=None,
+            header=None,
+            index_col=0)
+        df.index.name = 'YEAR'
+        df.columns = index
+
+        return df, self._extract_units(gases, units)
+
+    def _extract_units(self, gases, units):
+        combos = set(zip(gases, units))
+        result = {}
+        for v, u in combos:
+            if v not in result:
+                result[v] = u
+            else:
+                # this isn't expected to happen, but should check anyway
+                raise ValueError('Different units for {} in {}'.format(v, self.fname))
+
+        return result
+
+
+_file_types = {
+    'MAGICC6': MAGICC6Reader,
+    'MAGICC7': MAGICC7Reader,
+}
+
+
+def get_reader(fname):
+    with open(fname) as f:
+        lines = f.readlines()
+
+    # Infer the file type from the header
+    if '.__  __          _____ _____ _____ _____   ______   ______ __  __ _____  _____  _____ _   _' \
+            in lines[0]:
+        file_type = 'MAGICC7'
+    else:
+        file_type = 'MAGICC6'
+
+    return _file_types[file_type](fname, lines)
+
+
+class MAGICCInput(object):
+    """
+    An interface to read (and in future write) the input files used by MAGICC.
+
+    MAGICCInput can read input files from both MAGICC6 and MAGICC7. These
+    include files with extensions .IN and .SCEN7.
+
+    The MAGICCInput, once the target input file has been loaded it can be
```

Comment:
edit to become

'The MAGICCInput, once the target input file has been loaded, can be'